gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```diff
@@ -0,0 +1,178 @@
+
+  private ServerSocketChannel server;
+  private Thread thread;
+  private final Config config;
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    config = new Config(remoteHost, remotePort, proxyHost, proxyPort);
+  }
+
+  private Tunnel open() throws IOException {
+    try {
+      server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+      throw ioe;
+    }
+  }
+
+  public int getPort() throws IOException {
+    SocketAddress localAddress = null;
+
+    try {
+      if (server != null && server.isOpen()) {
+        localAddress = server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+      throw e;
+    }
+    return NON_EXISTENT_PORT;
+  }
+
+  private void startTunnelThread() {
+    thread = new Thread(new Dispatcher(), "Tunnel Listener");
+    thread.start();
+  }
+
+  public boolean isTunnelThreadAlive() {
+    return (thread != null && thread.isAlive());
+  }
+
+  private class Dispatcher implements Runnable {
+
+    private Selector selector;
+
+    public Dispatcher() {
+      try {
+        selector = Selector.open();
+      } catch (IOException e) {
+        LOG.error("Could not open selector",e);
+      }
+    }
+
+    @Override
+    public void run() {
+      try {
+        new AcceptHandler(server, selector, config);
+
+        while (!Thread.interrupted()) {
+
+          selector.select();
+          Set<SelectionKey> selectionKeys = selector.selectedKeys();
+
+          for (SelectionKey selectionKey : selectionKeys) {
+            dispatch(selectionKey);
+          }
+          selectionKeys.clear();
+        }
+      } catch (IOException ioe) {
+        LOG.error("Unhandled exception.  Tunnel will close", ioe);
+      }
+
+      LOG.info("Closing tunnel");
+    }
+
+    private void dispatch(SelectionKey selectionKey)
+        throws IOException {
+      Callable<?> attachment = (Callable<?>) selectionKey.attachment();
+
+      try {
+        attachment.call();
+      } catch (Exception e) {
+        LOG.warn("exception handling event on {}",selectionKey.channel(), e);
+      }
+    }
+  }
+
+  public void close() {
+    try {
+      thread.interrupt();
+      thread.join();
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      try {
+        server.close();
+        LOG.info("Closed tunnel.");
+      } catch (IOException ioe) {
+        LOG.warn("Failed to shutdown tunnel", ioe);
+      }
+    }
+  }
+
+  public static Tunnel build(String remoteHost, int remotePort, String proxyHost, int proxyPort)
+      throws IOException {
+    return new Tunnel(remoteHost, remotePort, proxyHost, proxyPort).open();
```

Comment:
True, right now, it's one Tunnel instance per JdbcProvider i.e. per JdbcExtractor. So it's a single server socket and a single thread per extractor instance, regardless of the number of connections, which off-hand does not seem too extravagant. I would say the biggest inefficiency is requiring two sockets per host (client -> Tunnel, Tunnel -> proxy) where one (client -> server) would do, but that just consumes file descriptors so it still seems affordable.

Agreed that it would be better to re-use the same Tunnel instance where possible, but that would lead to a large number of complications that may not be justified given the potential resource savings. We would have to implement some sort of coordination across multiple extractors so that they could share Tunnel instances where possible.

In our current job that ingests from 6 - 8 sources and a 100 different tables, if we have only up to 2 extractors, we would only have 2 Tunnels (with a server socket a thread each) for that job. My instinctive guess is that it should be OK with respect to resource usage, but maybe I need more information about the typical operating environment.